### ISSUE TITLE
print-missing option does not work with `nix build`

### ISSUE THREAD

**Issue by maralorn** [NixOS/nix#6561]
**Describe the bug**

The new experimental commands like `nix build` or `nix develop` do not call printMissing, even when `print-missing = true` (the default). 

**Steps To Reproduce**

1. Run `nix build --log-format raw` or `nix-build --log-format internal-json` for any package or flake.

**Expected behavior**

I would expect to see the list of packages which nix plans to build or download, since print-missing is set to true.

Alternatively, I would like the documentation of the print-missing option to state that it is not supposed to work with the new experimental commands, although that wouldn‘t solve my problem.

Note: When running `nix-build --log-format internal-json` the output of printMissing appears as a json message. So this is not an issue about the specific log-format, but about the difference between `nix-build` and `nix build`. The difference is, that in the default log-format of `nix build` the user wouldn’t notice the message, so they don’t miss it.

**`nix-env --version` output**

This happens on every nix version I know of.

**Additional context**

I am trying to make nix-output-monitor work with `nix build --log-format internal-json`. Sadly, when used with `nix build` the internal-json is missing a lot of information which the user could read from the output of nix-build. This makes figuring out the dependency tree which nix is planning to build impossible for nix-output-monitor.

**Comment by thufschmitt** [comment]
> The new experimental commands like nix build or nix develop do not call printMissing, even when print-missing = true (the default).

Yup' it doesn't indeed. `--dry-run` does though if that can help.

> I am trying to make nix-output-monitor work with nix build --log-format internal-json. Sadly, when used with nix build the internal-json is missing a lot of information which the user could read from the output of nix-build. This makes figuring out the dependency tree which nix is planning to build impossible for nix-output-monitor.

Which kind of information are you missing? I think this could be added to the logger (and that might be a better long-term choice since the output of `nix build` is likely to change in the future)

**Comment by maralorn** [comment]
With logger you mean the "internal-json" output or something else?

The information that nix-output-monitor needs, but currently cannot easily get from internal-json are in descending importance:

1. What are the targets of the current build, i.e. the derivations which don’t get build because they are depended on but because they are the end goal of the build? (It’s not important to know that they are the end goals, but I really need to know that they will be build because they are the roots of the build tree nix-output-monitor displays.)
2. What other derivations will be build?
3. What paths will be fetched?

If nix could improve the internal-json or another output method a little bit to give those information, that would be awesome!

**Comment by thufschmitt** [comment]
Yes sorry, I meant the internal-json one (well, any logger technically, but the json one is probably the most useful one in that case).

Ah sorry, I missed a dash and though that `nix build --log-format internal-json` was giving you that (but missing other stuff).

So yes, adding a call to `printMissing`  (probably just [here](https://github.com/thufschmitt/nix/blob/55cbd217069da5a60343bcaf7f265614167faf5b/src/libcmd/installables.cc#L843)) with a high-enough verbosity filter sounds right. And maybe also massage the output of `printMissing` to make it more json-esque.

Do you feel like getting a shot at it? Happy to give any guidance if you need

**Comment by maralorn** [comment]
> Ah sorry, I missed a dash and though that `nix build --log-format internal-json` was giving you that (but missing other stuff).

Those dashes be sneaky. :D

> So yes, adding a call to `printMissing` (probably just [here](https://github.com/thufschmitt/nix/blob/55cbd217069da5a60343bcaf7f265614167faf5b/src/libcmd/installables.cc#L843)) with a high-enough verbosity filter sounds right. And maybe also massage the output of `printMissing` to make it more json-esque.

Let’s phrase it this way: I would be very happy to receive the information in whatever way. For the current output of printMissing I actually already have a parser, so in the short term that would just mean more work. There are a lot of aspects/design decisions of the internal-json, which I don’t understand/probably could be improved/made more consistent. I probably could concoct a long wishlist for that. If I could discuss this with a nix developer, that might be very enlightening. Anyways I suggest we start with the minimal viable solution and then make our way from there. 

> Do you feel like getting a shot at it? Happy to give any guidance if you need

I mean, the whole point of nix-output-monitor is that I didn‘t want to touch C++ anymore.^^ But sure I can give it a shot.
